import_export/admin.py

```diff
@@ -213,6 +211,18 @@ def get_import_data_kwargs(self, request, *args, **kwargs):
             return kwargs
         return {}
 
+    def read_from_tmp_storage(self, tmp_storage, input_format):
+        if input_format.is_binary():
+            data = tmp_storage.read(input_format.get_read_mode())
+        else:
+            data = tmp_storage.read(input_format.get_read_mode(), encoding=self.from_encoding)
+
+            # if the object is being read from a cache, then it might be binary
+            # if so, decode before returning
+            if isinstance(data, bytes):
+                data = force_str(data, encoding=self.from_encoding)
```

Comment:
This is what is keeping me from approving this PR. If you ask the `tmp_storage` to read the contents with an encoding, why do you need to check for bytes and decode again? It looks to me that this should be done in the tmp storage class. If this only applies to CacheStorage then move it there. What do you think?

Reply:
MediaStorage would also need this check as the django Storage class open method does not include a encoding parameter. So the changes I'm talking about are basically on the [PR](https://github.com/django-import-export/django-import-export/pull/1316/files) along with some test :)

Reply:
Thanks @manelclos.  I've made some changes:

- Adjusted `tmp_storages.py` so that the interface is unchanged (`encoding` now an optional constructor param)
- Added decoding for binary encoded files
- Removed decoding from `admin.py`
- Added tests

Let me know what you think

Reply:
Looks great now, thanks! `MediaStorage` should also check for bytes and apply encoding, right?

Reply:
`MediaStorage` is backed by `django.core.files.storage.Storage`, which doesn't accept the `encoding` arg, hence we cannot apply the encoding for this.

Reply:
My understanding is that the read call will return the file contents, here: https://github.com/django-import-export/django-import-export/pull/1306/files#diff-08b7301345c3436500f8df072dcf1f6e7592ab9dda9a5620e8ed5d3c8252f493R84-R86

So we can save the returned content in a data variable, check it for binary and encoding, basically what we do for the CacheStorage, here: https://github.com/django-import-export/django-import-export/pull/1306/files#diff-08b7301345c3436500f8df072dcf1f6e7592ab9dda9a5620e8ed5d3c8252f493L66-R70

Or maybe I'm missing something :)

Reply:
In the case of `MediaStorage`, it's not necessary to decode, because the `save()` operation saves it correctly in its encoded form (this happens in the save operation in [ContentFile](https://docs.djangoproject.com/en/4.0/ref/files/file/#the-contentfile-class)).  The tests in `test_admin_integration.py` prove this works ok.  I've added a comment in `MediaStorage` to clarify.

Reply:
Hi @matthewhegarty, we need to fix the tests then ;)

If I upload a file with ISO-8859-1 encoding, the `ContentFile` will save the bytes (as in character codes) it gets, as there has been no decoding, no magic here. The linked documentation will use unicode => utf8 to save that Spanish text.

While testing that assumption using the core app, I found that we need to force reading in binary mode when the encoding is set, otherwise the read operation will fail.

It should be something like this:
```python
class MediaStorage(BaseStorage):

    def read(self, read_mode='rb'):
        if self.encoding is not None:
            read_mode = 'rb'
        with default_storage.open(self.get_full_path(), mode=read_mode) as f:
            data = f.read()
            if isinstance(data, bytes) and self.encoding is not None:
                data = data.decode(self.encoding)
            return data
```

Please let me know if this makes sense.

Reply:
@manelclos I've made quite a few changes, along the lines we discussed.  This is quite a tricky problem!  See what you think.

Reply:
@manelclos I think this is ready to go now - are you able to take another look and let me know what you think?